design/006-renku-api-activities-plans/006-renku-api-activities-plans.md

```diff
@@ -0,0 +1,488 @@
+activities_by_filtered_parameter = Activity.filter_py_parameter(name="lr", value=lambda v: v > 0.5)
+```
+
+`.list()` returns all activities in the project.
+
+`.filter_by_input()` and `.filter_by_output()` return all activities
+that used a certain input or produced a certain output. Setting
+`with_children=True` means that if a directory is passed, all activities for
+that directory and its children are returned.
+
+`.filter_py_parameter()` filters activities based on parameters' names and/or
+values.
+
+Parameters `path`, `name` and `value` should accept plain values, list of
+values or a `Callable` accepting the value and returning a `bool`, to filter
+on a single value, list of values or custom condition, respectively.
+
+```Python
+from renku.api import Plan
+
+all_active_plans = Plan.list()
+all_plans = Plan.list(include_deleted=True)
+```
+
+`.list()` returns all plans in the project. `include_deleted=True` would
+also return plans that were deleted.
+
+```Python
+from renku.api import Project
+
+status = Project().status()
+status = Project().status(outputs=['my_output'])
+```
+
+`.status()` returns the current project status (`ProjectStatus`), detailing
+detailing stale outputs and activities as well as modified and deleted inputs
+that cause them to be stale. With `outputs` specified, limits status to those
+outputs and activities producing them.
+
+### Viewing the Inputs, Outputs, and Parameters of Plans and Activities
+To see the names and default values of all the input, output, or parameters fields in a Plan:
+
+```python
+from renku.api import Plan
+
+all_active_plans = Plan.list()
+my_plan = all_active_plans[0]
+
+input_fields_list = my_plan.input_fields
+output_fields_list = my_plan.output_fields
+parameter_fields_list = my_plan.parameter_fields
+
+# print Input fields and their default values
+for input_field in input_fields_list:
+    print(f"{input_field.name}: {input_field.value})
+```
+
+
+
+And to see the input or output filepaths associated with an activity:
+
+```python
+from renku.api import Activity
+
+all_activities = Activity.list()
+my_activity = all_activities[0]
+
+input_files = my_activity.inputs
+output_files = my_activity.outputs
+for f in input_files:
+    print(f"{f.path}")
+
+parameters = my_Activity.parameters
+for p in parameters:
+    print(f"{p.field.name}: {p.value}")
+
+```
+
+To see the actual values that were specified for the input and output fields at the
+execution of the activity (the input & output _field values_ may be different from
+the input and output _paths_ if the fields were interpretted. For example, if an
+`OutputField` was set to `my_file_{yyyy-mm-dd}.txt`, then the value in the
+`Outputs` list would be `my_file_2022-03-30.txt`):
+
+```python
+from renku.api import Activity
+
+all_activities = Activity.list()
+my_activity = all_activities[0]
+executed_plan = my_activity.executed_plan
+# ??? what is the return value of `Activity.executed_plan`?
```

Comment:
```suggestion
executed_plan = my_activity.executed_plan
base_plan = my_activity.base_plan
for i, ex_pf in enumerate(executed_plan.parameter_fields):
    base_pf = base_plan.parameter_fields[i]
    print("Plan executed with the following parameters (default in parens)"
    print(f"{exec_pf.value} ({base_rf.value)}")
```
Re executed plans- is this how it would work to compare base vs executed plan parameters?

Reply:
yes, exactly (plus the same for input fields and output fields).

In our codebase this corresponds to:

```
executed_plan = my_activity.plan_with_values

for i, p in enumerate(executed_plan.parameters):
    print("Plan executed with the following parameters (default in parens)"
    print(f"{p.actual_value} ({p.default_value)}")
```

Reply:
I've replaced it with just `executed_command` as discussed over lunch

Reply:
🌟